### PR TITLE
Do not allow agent to register with duplicate elastic agent id (#3828)

### PR DESCRIPTION
**Pull request by GaneshSPatil**
[Agent Registration] Do not allow agent registration if agent with same elastic agent id already exists
[XSL Migration] Add a xslt transformation to remove an agent with duplicate elastic agent id.

### CODE REVIEW

File: config/config-server/resources/upgrades/103.xsl [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        <xsl:attribute name="schemaVersion">103</xsl:attribute>
+    </xsl:template>
+    <!-- Copy everything -->
+    <xsl:template match="@*|node()">
+        <xsl:copy>
+            <xsl:apply-templates select="@*|node()"/>
+        </xsl:copy>
+    </xsl:template>
+
+    <xsl:template match="agent[@elasticAgentId = preceding-sibling::agent/@elasticAgentId]"/>
```
Comment (ketan) [inline config/config-server/resources/upgrades/103.xsl]: What if the preceeding sibling is not the dup, but some other sibling is?

Comment (bdpiparva) [reply config/config-server/resources/upgrades/103.xsl]: @ketan --- It think will check for all siblings. see http://xsltransform.net/3MEbY6h/3.

File: server/test/integration/com/thoughtworks/go/config/GoConfigMigrationIntegrationTest.java [hunk 2]
```diff
@@ -1854,6 +1854,22 @@ public void shouldSkipParamResoulutionForElasticConfig_asPartOf100To101Migration
         );
     }
 
+
+    @Test
+    public void shouldRemoveAgentWithDuplicateElasticAgentId_asPartOf102To103Migration() throws NoSuchMethodException, IllegalAccessException, InvocationTargetException {
+        String configXml = "<cruise schemaVersion='102'>" +
+                "<agents>\n" +
+                "    <agent hostname=\"hostname\" ipaddress=\"127.0.0.1\" uuid=\"c46a08a7-921c-4e77-b748-6128975a3e7d\" elasticAgentId=\"16649813-4cb3-4682-8702-8e202824dd73\" elasticPluginId=\"elastic-plugin-id\" />\n" +
```
Comment (bdpiparva) [inline server/test/integration/com/thoughtworks/go/config/GoConfigMigrationIntegrationTest.java]: @GaneshSPatil --- Change the order of duplicate agents. Just to make sure that translation will remove duplicates in any order.

File: server/src/com/thoughtworks/go/server/controller/AgentRegistrationController.java [hunk 3]
```diff
@@ -233,6 +233,10 @@ public ResponseEntity agentRequest(@RequestParam("hostname") String hostname,
                 return new ResponseEntity<>("Elastic agents must submit both elasticAgentId and elasticPluginId.", UNPROCESSABLE_ENTITY);
             }
 
+            if (elasticAgentIdAlreadyRegistered(elasticAgentId, elasticPluginId)) {
```
Comment (bdpiparva) [inline server/src/com/thoughtworks/go/server/controller/AgentRegistrationController.java]: Can we move this check inside [AgentConfig#validate](https://github.com/gocd/gocd/blob/master/config/config-api/src/com/thoughtworks/go/config/AgentConfig.java#L83-L111) ?

**Pull request merged**